agent/agentcontainers/containers_dockercli.go

```diff
@@ -367,50 +371,80 @@ func (dis dockerInspectState) String() string {
 	return sb.String()
 }
 
-func convertDockerInspect(in dockerInspect) (codersdk.WorkspaceAgentDevcontainer, []string) {
+func convertDockerInspect(raw string) ([]codersdk.WorkspaceAgentDevcontainer, []string, error) {
 	var warns []string
-	out := codersdk.WorkspaceAgentDevcontainer{
-		CreatedAt: in.Created,
-		// Remove the leading slash from the container name
-		FriendlyName: strings.TrimPrefix(in.Name, "/"),
-		ID:           in.ID,
-		Image:        in.Config.Image,
-		Labels:       in.Config.Labels,
-		Ports:        make([]codersdk.WorkspaceAgentListeningPort, 0),
-		Running:      in.State.Running,
-		Status:       in.State.String(),
-		Volumes:      make(map[string]string, len(in.Mounts)),
-	}
-
-	if in.HostConfig.PortBindings == nil {
-		in.HostConfig.PortBindings = make(map[string]any)
-	}
-	portKeys := maps.Keys(in.HostConfig.PortBindings)
-	// Sort the ports for deterministic output.
-	sort.Strings(portKeys)
-	for _, p := range portKeys {
-		if port, network, err := convertDockerPort(p); err != nil {
-			warns = append(warns, err.Error())
-		} else {
-			out.Ports = append(out.Ports, codersdk.WorkspaceAgentListeningPort{
-				Network: network,
-				Port:    port,
-			})
-		}
+	var ins []dockerInspect
+	if err := json.NewDecoder(strings.NewReader(raw)).Decode(&ins); err != nil {
+		return nil, nil, xerrors.Errorf("decode docker inspect output: %w", err)
 	}
+	outs := make([]codersdk.WorkspaceAgentDevcontainer, 0, len(ins))
 
-	if in.Mounts == nil {
-		in.Mounts = []dockerInspectMount{}
-	}
-	// Sort the mounts for deterministic output.
-	sort.Slice(in.Mounts, func(i, j int) bool {
-		return in.Mounts[i].Source < in.Mounts[j].Source
-	})
-	for _, k := range in.Mounts {
-		out.Volumes[k.Source] = k.Destination
+	for _, in := range ins {
+		out := codersdk.WorkspaceAgentDevcontainer{
+			CreatedAt: in.Created,
+			// Remove the leading slash from the container name
+			FriendlyName: strings.TrimPrefix(in.Name, "/"),
+			ID:           in.ID,
+			Image:        in.Config.Image,
+			Labels:       in.Config.Labels,
+			Ports:        make([]codersdk.WorkspaceAgentListeningPort, 0),
+			Running:      in.State.Running,
+			Status:       in.State.String(),
+			Volumes:      make(map[string]string, len(in.Mounts)),
+		}
+
+		if in.NetworkSettings.Ports == nil {
+			in.NetworkSettings.Ports = make(map[string][]dockerInspectPort)
+		}
+		portKeys := maps.Keys(in.NetworkSettings.Ports)
+		// Sort the ports for deterministic output.
+		sort.Strings(portKeys)
+		// Each port binding may have multiple entries mapped to the same interface.
+		// Keep track of the ports we've already seen.
```

Comment:
Nit: This comment could be clarified that we only care about which ports are exposed on the host, ignoring the mapping between host and container ports.

(Here I echo the concern I mentioned in my previous review: https://github.com/coder/coder/pull/16866#pullrequestreview-2675114497)

Reply:
Updated!